src/irr/asset/CImageLoaderTGA.cpp

```diff
@@ -112,78 +112,129 @@ bool CImageLoaderTGA::isALoadableFileFormat(io::IReadFile* _file) const
+		newRegions->front() = *referenceRegion;
+
+		newImageParams.format = destFormat;
+		newConvertedImage = ICPUImage::create(std::move(newImageParams));
+		newConvertedImage->setBufferAndRegions(std::move(newCpuBuffer), newRegions);
+	}
+
+	auto attachedRegion = newConvertedImage->getRegions().begin();
+
+	state.inImage = inputCreationImage.get();
+	state.outImage = newConvertedImage.get();
+	state.inOffset = { 0, 0, 0 };
+	state.inBaseLayer = 0;
+	state.outOffset = { 0, 0, 0 };
+	state.outBaseLayer = 0;
+	state.extent = attachedRegion->getExtent();
+	state.layerCount = attachedRegion->imageSubresource.layerCount;
+	state.inMipLevel = attachedRegion->imageSubresource.mipLevel;
+	state.outMipLevel = attachedRegion->imageSubresource.mipLevel;
+
+	if (!convertFilter.execute(&state))
+		os::Printer::log("Something went wrong while converting!", ELL_WARNING);
+
+	bool OpenGlFlip = !flip;
+	if (OpenGlFlip)
+	{
+		auto newFormat = newConvertedImage->getCreationParameters().format;
+		asset::TexelBlockInfo blockInfo(newFormat);
+		core::vector3du32_SIMD trueExtent = blockInfo.convertTexelsToBlocks(newConvertedImage->getRegions().begin()->getTexelStrides());
+
+		auto buffer = newConvertedImage->getBuffer();
+		auto out = reinterpret_cast<uint8_t*>(buffer->getPointer()) + buffer->getSize();
+		auto stride = trueExtent.X * getTexelOrBlockBytesize(newFormat);
+
+		for (uint32_t y = 0; y < trueExtent.Y; ++y)
 			out -= stride;
-		
-		const void *src_container[4] = {in, nullptr, nullptr, nullptr};
-		convertColor<srcFormat, destFormat>(src_container, out, stride, size);
-		in += stride;
-		
-		if (!flip)
-			out += stride;
 	}
-}
+
+	return newConvertedImage;
+};
 
 //! creates a surface from the file
 asset::SAssetBundle CImageLoaderTGA::loadAsset(io::IReadFile* _file, const asset::IAssetLoader::SAssetLoadParams& _params, asset::IAssetLoader::IAssetLoaderOverride* _override, uint32_t _hierarchyLevel)
 {
 	STGAHeader header;
-
-	core::smart_refctd_ptr<ICPUBuffer> palette = nullptr;
-
 	_file->read(&header, sizeof(STGAHeader));
 
+	core::smart_refctd_ptr<ICPUImage> colorMap; // not used, but it's texel buffer may be useful in future
 	const auto bytesPerTexel = header.PixelDepth / 8;
 
-	// skip image identification field
-	if (header.IdLength)
+	if (header.IdLength) // skip image identification field
 		_file->seek(header.IdLength, true);
 
 	if (header.ColorMapType)
 	{
-		// create 32 bit palette
-		palette = core::make_smart_refctd_ptr<ICPUBuffer>(header.ColorMapLength * sizeof(uint32_t));
-
-		// read color map
-		uint8_t* colorMap = _IRR_NEW_ARRAY(uint8_t, header.ColorMapEntrySize / 8 * header.ColorMapLength);
-		_file->read(colorMap,header.ColorMapEntrySize/8 * header.ColorMapLength);
+		auto colorMapEntryByteSize = header.ColorMapEntrySize / 8 * header.ColorMapLength;
+		uint8_t* colorMapEntry = _IRR_NEW_ARRAY(uint8_t, colorMapEntryByteSize);
```

Comment:
you could have that be an `ICPUBuffer` already

Reply:
outdated